### PR TITLE
test(smartplaylists): add smart playlist e2e test suite

### PR DESCRIPTION
**Pull request by deluan**
### Description

This PR adds a comprehensive end-to-end test suite for smart playlists (NSP). The suite populates a real SQLite database by scanning a FakeFS with 8 classic rock tracks, then verifies criteria evaluation across all field types.

The test architecture mirrors the existing Subsonic e2e pattern (`server/e2e/`): scan once in `BeforeSuite`, snapshot the DB, restore from snapshot in each `BeforeEach` for full isolation.

**42 test cases** covering:
- **String fields** (5): `is`, `isNot`, `contains`, `startsWith`, `endsWith`
- **Numeric fields** (3): `gt`, `lt`, `inTheRange` for year and BPM
- **Boolean fields** (2): compilation true/false
- **File type** (1): filetype matching (mp3/flac)
- **Multi-valued tags** (4): genre `is`, `isNot`, `contains` across multi-valued entries
- **Participants** (3): composer `is`, `contains`, `isNot` across slash-separated values
- **Annotations** (6): starred, rating, play count with various operators
- **Date/time fields** (5): `inTheLast`, `notInTheLast`, `after` for lastPlayed, dateLoved, dateRated, dateAdded
- **Logic operators** (3): `ALL` (AND), `ANY` (OR), nested all/any
- **Sorting & limits** (3): ordered results, multi-field sort, combined real-world pattern
- **Playlist operators** (7): `inPlaylist`/`notInPlaylist` with public/private visibility, cross-user ownership, and recursive smart playlist refresh

### Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Refactor
- [ ] Other (please describe):

### Checklist

- [x] My code follows the project's coding style
- [x] I have tested the changes locally
- [ ] I have added or updated documentation as needed
- [x] I have added tests that prove my fix/feature works (or explain why not)
- [x] All existing and new tests pass

### How to Test

```bash
make test PKG=./core/playlists/e2e
```

All 42 specs should pass. The suite takes ~1 second.

To add a new edge case test, add an `It()` block with an `evaluateRule()` call:
```go
It("describes the edge case", func() {
    results := evaluateRule(`{"all":[{"is":{"genre":"Blues"}}]}`)
    Expect(results).To(ConsistOf("Come Together", "Black Dog", "All Along the Watchtower"))
})
```

### Additional Notes

- The suite is fully isolated in `core/playlists/e2e/` — no shared fixtures with existing persistence tests
- Annotation criteria (rating, playcount) use JSON numbers instead of strings to avoid SQLite type affinity issues with COALESCE expressions
- Date/time tests for `inTheLast`/`notInTheLast` rely on annotations being set during the same test run (today), which is reliable since BeforeSuite always runs immediately before the specs

### CODE REVIEW

File: core/playlists/e2e/e2e_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+	loaded, err := ds.Playlist(ctx).GetWithTracks(pls.ID, true, false)
+	Expect(err).ToNot(HaveOccurred())
+
+	titles := make([]string, len(loaded.Tracks))
+	for i, t := range loaded.Tracks {
+		titles[i] = t.Title
+	}
+	return titles
+}
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/e2e_suite_test.go:176]: Fixed. `evaluateRule` now delegates to `evaluateRuleOrdered` + `sort.Strings`.

File: core/playlists/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	rows, err := sqlDB.Query("SELECT name FROM main.sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' AND name NOT LIKE '%_fts' AND name NOT LIKE '%_fts_%'")
+	Expect(err).ToNot(HaveOccurred())
+	var tables []string
+	for rows.Next() {
+		var name string
+		Expect(rows.Scan(&name)).To(Succeed())
+		tables = append(tables, name)
+	}
+	Expect(rows.Err()).ToNot(HaveOccurred())
+	rows.Close()
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/e2e_suite_test.go]: Fixed. Table list is now cached in `snapshotTables` during `BeforeSuite` instead of re-querying per test.

File: core/playlists/e2e/e2e_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+func restoreDB() {
+	sqlDB := db.Db()
+
+	_, err := sqlDB.Exec("PRAGMA foreign_keys = OFF")
+	Expect(err).ToNot(HaveOccurred())
+
+	_, err = sqlDB.Exec("ATTACH DATABASE ? AS snapshot", snapshotPath)
+	Expect(err).ToNot(HaveOccurred())
+
+	rows, err := sqlDB.Query("SELECT name FROM main.sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' AND name NOT LIKE '%_fts' AND name NOT LIKE '%_fts_%'")
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/e2e_suite_test.go]: Fixed. Using `defer rows.Close()` now. Also moved the query to `BeforeSuite` so it only runs once.

File: core/playlists/e2e/e2e_suite_test.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+	Expect(rows.Err()).ToNot(HaveOccurred())
+	rows.Close()
+
+	for _, table := range tables {
+		// Table names come from sqlite_master, not user input, so concatenation is safe here
+		_, err = sqlDB.Exec(`DELETE FROM main."` + table + `"`) //nolint:gosec
+		Expect(err).ToNot(HaveOccurred())
+		_, err = sqlDB.Exec(`INSERT INTO main."` + table + `" SELECT * FROM snapshot."` + table + `"`) //nolint:gosec
+		Expect(err).ToNot(HaveOccurred())
+	}
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/e2e_suite_test.go]: Fixed. `restoreDB` now uses BEGIN/COMMIT with a deferred ROLLBACK as safety net, plus defers for DETACH and foreign_keys re-enable.

File: core/playlists/e2e/smartplaylist_test.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+			Expect(results).To(ConsistOf("Stairway To Heaven", "Black Dog", "Bohemian Rhapsody", "We Are the Champions"))
+		})
+
+		It("matches by year less than", func() {
+			results := evaluateRule(`{"all":[{"lt":{"year":"1969"}}]}`)
+			Expect(results).To(ConsistOf("So What", "All Along the Watchtower"))
+		})
+
+		It("matches by BPM in range", func() {
+			results := evaluateRule(`{"all":[{"inTheRange":{"bpm":["100","130"]}}]}`)
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/smartplaylist_test.go]: Fixed. All numeric criteria values now use JSON numbers to match the canonical criteria JSON shape.

File: core/playlists/e2e/smartplaylist_test.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+	Describe("Date/time fields", func() {
+		It("matches lastPlayed inTheLast 1 day", func() {
+			results := evaluateRule(`{"all":[{"inTheLast":{"lastplayed":"1"}}]}`)
+			Expect(results).To(ConsistOf("Come Together", "Black Dog"))
+		})
+
+		It("matches lastPlayed notInTheLast (far future)", func() {
+			results := evaluateRule(`{"all":[{"notInTheLast":{"lastplayed":"99999"}}]}`)
+			Expect(results).To(ConsistOf("Something", "Stairway To Heaven", "So What",
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/smartplaylist_test.go:179]: Fixed. Period values for `inTheLast`/`notInTheLast` now use JSON numbers.

File: core/playlists/e2e/e2e_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+	_, err = sqlDB.Exec("DETACH DATABASE snapshot")
+	Expect(err).ToNot(HaveOccurred())
+	_, err = sqlDB.Exec("PRAGMA foreign_keys = ON")
+	Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (deluan) [reply core/playlists/e2e/e2e_suite_test.go:314]: Fixed. `restoreDB` now uses defers for DETACH, foreign_keys, and a safety ROLLBACK.
